File: web/src/services/voiceService.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import {
  createDraftTaskFromExtracted,
  switchDraftTimedPlanType,
  sendVoiceCapture,
  type ExtractedDraftPayload,
  type DraftTask,
} from "./voiceService";
import type { Task } from "../contracts/task";
import type { SupabaseClient } from "@supabase/supabase-js";
import { formatPlanDate, formatPlanTime } from "./temporalService";

describe("Voice Service - Draft Task Creation & Plan Semantics", () => {
  it("creates a Date-only Draft when only plan_date is present (no type)", () => {
    const payload: ExtractedDraftPayload = {
      title: "Buy groceries",
      description: "Apples, bananas, milk",
      priority: 2,
      plan_date: "2026-08-25",
      plan_time: null,
      plan_type: null,
    };

    const draft = createDraftTaskFromExtracted(payload, "instant");

    expect(draft.title).toBe("Buy groceries");
    expect(draft.description).toBe("Apples, bananas, milk");
    expect(draft.priority).toBe(2);
    expect(draft.plan).toEqual({ date: "2026-08-25" });
    expect(draft.validationError).toBeNull();
  });

  it("applies effective default (Instant) when plan has date and time but unstated type", () => {
    const payload: ExtractedDraftPayload = {
      title: "Team standup",
      plan_date: "2026-08-25",
      plan_time: "09:30:00",
      plan_type: null,
    };

    const draft = createDraftTaskFromExtracted(payload, "instant");

    expect(draft.plan).not.toBeNull();
    expect(draft.plan?.type).toBe("instant");
    if (draft.plan && "at" in draft.plan) {
      expect(draft.plan.at).toBeDefined();
    }
    expect(draft.validationError).toBeNull();
  });

  it("applies effective default (Floating) when plan has date and time and default is floating", () => {
    const payload: ExtractedDraftPayload = {
      title: "Team standup",
      plan_date: "2026-08-25",
      plan_time: "09:30:00",
      plan_type: null,
    };

    const draft = createDraftTaskFromExtracted(payload, "floating");

    expect(draft.plan).toEqual({
      type: "floating",
      date: "2026-08-25",
      time: "09:30",
    });
    expect(draft.validationError).toBeNull();
  });

  it("honors explicit spoken Instant type overriding Floating default", () => {
    const payload: ExtractedDraftPayload = {
      title: "Flight departure",
      plan_date: "2026-08-25",
      plan_time: "14:00:00",
      plan_type: "instant",
    };

    const draft = createDraftTaskFromExtracted(payload, "floating");

    expect(draft.plan?.type).toBe("instant");
    if (draft.plan && "at" in draft.plan) {
      expect(draft.plan.at).toBeDefined();
    }
  });

  it("flags validation error when speech explicitly requests Instant/Floating without a clock time", () => {
    const payload: ExtractedDraftPayload = {
      title: "Invalid explicit floating task",
      plan_date: "2026-08-25",
      plan_time: null,
      plan_type: "floating",
    };

    const draft = createDraftTaskFromExtracted(payload, "instant");

    expect(draft.validationError).toBe(
      "An explicit Instant or Floating plan requires a clock time. Please provide a time or change to Date-only.",
    );
    expect(draft.plan).toEqual({ date: "2026-08-25" });
  });

  it("switching Instant <-> Floating preserves displayed calendar date and clock time", () => {
    const initialPayload: ExtractedDraftPayload = {
      title: "Preserve displayed time test",
      plan_date: "2026-08-25",
      plan_time: "19:00:00",
      plan_type: "instant",
    };

    const draftInstant = createDraftTaskFromExtracted(
      initialPayload,
      "instant",
    );
    expect(draftInstant.plan?.type).toBe("instant");
    const displayedDateBefore = formatPlanDate(draftInstant.plan);
    const displayedTimeBefore = formatPlanTime(draftInstant.plan);

    const draftFloating = switchDraftTimedPlanType(
      draftInstant,
      "floating",
      "instant",
    );
    expect(draftFloating.plan?.type).toBe("floating");
    expect(formatPlanDate(draftFloating.plan)).toBe(displayedDateBefore);
    expect(formatPlanTime(draftFloating.plan)).toBe(displayedTimeBefore);

    const draftInstantAgain = switchDraftTimedPlanType(
      draftFloating,
      "instant",
      "instant",
    );
    expect(draftInstantAgain.plan?.type).toBe("instant");
    expect(formatPlanDate(draftInstantAgain.plan)).toBe(displayedDateBefore);
    expect(formatPlanTime(draftInstantAgain.plan)).toBe(displayedTimeBefore);
  });
});

describe("Voice Service - API & Error Handling", () => {
  let mockSupabaseClient: SupabaseClient;
  let mockFetch: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    mockSupabaseClient = {
      supabaseUrl: "https://example.supabase.co",
      auth: {
        getSession: vi.fn().mockResolvedValue({
          data: {
            session: { access_token: "mock-valid-token" },
          },
        }),
      },
      from: vi.fn(),
    } as unknown as SupabaseClient;

    mockFetch = vi.fn();
    (globalThis as unknown as { fetch: typeof fetch }).fetch =
      mockFetch as unknown as typeof fetch;
  });

  it("handles successful multi-draft voice create response", async () => {
    mockFetch.mockResolvedValue({
      ok: true,
      json: async () => ({
        transcript: "Buy milk tomorrow and call doctor at 3pm",
        mode: "create",
        drafts: [
          {
            title: "Buy milk",
            description: null,
            priority: 4,
            plan_date: "2026-08-22",
            plan_time: null,
            plan_type: null,
          },
          {
            title: "Call doctor",
            description: null,
            priority: 2,
            plan_date: "2026-08-21",
            plan_time: "15:00:00",
            plan_type: null,
          },
        ],
      }),
    });

    const result = await sendVoiceCapture(mockSupabaseClient, {
      audioBlob: new Blob(["mock-wav"], { type: "audio/wav" }),
      recordingStartTime: "2026-08-21T10:00:00Z",
      timezone: "UTC",
      effectiveDefaultTimedPlanType: "instant",
    });

    expect(result.transcript).toBe(
      "Buy milk tomorrow and call doctor at 3pm",
    );
    expect(result.drafts).toHaveLength(2);
    expect(result.drafts[0].title).toBe("Buy milk");
    expect(result.drafts[0].plan).toEqual({ date: "2026-08-22" });
    expect(result.drafts[1].title).toBe("Call doctor");
    expect(result.drafts[1].priority).toBe(2);
    expect(result.drafts[1].plan?.type).toBe("instant");
  });

  it("handles voice edit when focused task is provided, preserving existing timed type unless changed", async () => {
    const existingTask: Task = {
      id: "task-1234",
      title: "Existing Task",
      description: "Original description",
      priority: 3,
      plan: {
        type: "floating",
        date: "2026-08-20",
        time: "10:00:00",
      },
      labels: [],
      parentId: null,
      completedAt: null,
      createdAt: "2026-08-20T00:00:00Z",
      updatedAt: "2026-08-20T00:00:00Z",
      version: 1,
    };

    mockFetch.mockResolvedValue({
      ok: true,
      json: async () => ({
        transcript: "Move to tomorrow at 11am",
        mode: "edit",
        edit: {
          title: null,
          description: null,
          priority: null,
          plan_date: "2026-08-22",
          plan_time: "11:00:00",
          plan_type: null, // Did not explicitly change type -> preserve floating!
          clear_plan: null,
        },
      }),
    });

    const result = await sendVoiceCapture(mockSupabaseClient, {
      audioBlob: new Blob(["mock-wav"], { type: "audio/wav" }),
      recordingStartTime: "2026-08-21T10:00:00Z",
      timezone: "UTC",
      focusedTask: existingTask,
      effectiveDefaultTimedPlanType: "instant",
    });

    expect(result.mode).toBe("edit");
    expect(result.drafts).toHaveLength(1);
    const editDraft = result.drafts[0];
    expect(editDraft.originalTaskId).toBe("task-1234");
    expect(editDraft.title).toBe("Existing Task");
    expect(editDraft.plan?.type).toBe("floating"); // Preserved floating!
    expect(
      editDraft.plan && "time" in editDraft.plan ? editDraft.plan.time : null,
    ).toBe("11:00");
  });

  it("handles voice correction across existing drafts", async () => {
    const existingDrafts: DraftTask[] = [
      {
        id: "draft-1",
        title: "Buy milk",
        description: null,
        priority: 4,
        plan: { date: "2026-08-22" },
        labels: [],
        parentId: null,
        validationError: null,
      },
      {
        id: "draft-2",
        title: "Go to pool",
        description: null,
        priority: 4,
        plan: { date: "2026-08-25" },
        labels: [],
        parentId: null,
        validationError: null,
      },
    ];

    mockFetch.mockResolvedValue({
      ok: true,
      json: async () => ({
        transcript: "No, buy milk today, not tomorrow",
        mode: "create",
        drafts: [
          {
            target_draft_index: 0,
            title: "Buy milk",
            description: null,
            priority: 4,
            plan_date: "2026-08-21",
            plan_time: null,
            plan_type: null,
          },
        ],
      }),
    });

    const result = await sendVoiceCapture(mockSupabaseClient, {
      audioBlob: new Blob(["mock-wav"], { type: "audio/wav" }),
      recordingStartTime: "2026-08-21T10:00:00Z",
      timezone: "UTC",
      existingDrafts,
      effectiveDefaultTimedPlanType: "instant",
    });

    expect(result.drafts).toHaveLength(2);
    expect(result.drafts[0].title).toBe("Buy milk");
    expect(result.drafts[0].plan).toEqual({ date: "2026-08-21" }); // updated!
    expect(result.drafts[1].title).toBe("Go to pool");
    expect(result.drafts[1].plan).toEqual({ date: "2026-08-25" }); // untouched!
  });

  it("throws VoiceError on 429 rate limit with earliest retry timestamp", async () => {
    mockFetch.mockResolvedValue({
      ok: false,
      status: 429,
      json: async () => ({
        error: "Rate limit exceeded: maximum 3 requests per minute.",
        code: "rate_limit_minute",
        earliest_retry_at: "2026-08-21T10:01:00Z",
        retry_after_seconds: 45,
      }),
    });

    await expect(
      sendVoiceCapture(mockSupabaseClient, {
        audioBlob: new Blob(["mock-wav"], { type: "audio/wav" }),
        recordingStartTime: "2026-08-21T10:00:00Z",
        timezone: "UTC",
        effectiveDefaultTimedPlanType: "instant",
      }),
    ).rejects.toMatchObject({
      status: 429,
      code: "rate_limit_minute",
      earliestRetryAt: "2026-08-21T10:01:00Z",
      retryAfterSeconds: 45,
    });
  });

  it("throws VoiceError on 503 circuit breaker", async () => {
    mockFetch.mockResolvedValue({
      ok: false,
      status: 503,
      json: async () => ({
        error: "Voice capture is temporarily unavailable. Please try again later.",
        code: "circuit_breaker_daily",
      }),
    });

    await expect(
      sendVoiceCapture(mockSupabaseClient, {
        audioBlob: new Blob(["mock-wav"], { type: "audio/wav" }),
        recordingStartTime: "2026-08-21T10:00:00Z",
        timezone: "UTC",
        effectiveDefaultTimedPlanType: "instant",
      }),
    ).rejects.toMatchObject({
      status: 503,
      code: "circuit_breaker_daily",
    });
  });

  it("flags isNetworkError when network fetch fails so local recording is preserved", async () => {
    mockFetch.mockRejectedValue(new TypeError("Failed to fetch"));

    await expect(
      sendVoiceCapture(mockSupabaseClient, {
        audioBlob: new Blob(["mock-wav"], { type: "audio/wav" }),
        recordingStartTime: "2026-08-21T10:00:00Z",
        timezone: "UTC",
        effectiveDefaultTimedPlanType: "instant",
      }),
    ).rejects.toMatchObject({
      isNetworkError: true,
      code: "network_error",
    });
  });
});

File: android/app/src/test/java/com/cras/app/voice/VoiceServiceTest.kt
package com.cras.app.voice

import com.cras.app.auth.OperatorSession
import com.cras.app.config.PublicSupabaseConfig
import com.cras.app.domain.TimedPlanType
import com.cras.app.models.Plan
import com.cras.app.models.Task
import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.util.UUID
import java.util.concurrent.TimeUnit

private fun session() = OperatorSession(
    operatorId = UUID.randomUUID().toString(),
    email = "op@example.com",
    accessToken = "test-access-token",
)

private fun task(
    id: String = UUID.randomUUID().toString(),
    title: String = "Existing Task",
    description: String? = "Original description",
    priority: Int = 3,
    plan: Plan? = Plan.Floating(date = "2026-08-20", time = "10:00"),
): Task = Task(
    id = id,
    title = title,
    description = description,
    priority = priority,
    plan = plan,
    labels = emptyList(),
    parentId = null,
    completedAt = null,
    createdAt = "2026-08-20T00:00:00Z",
    updatedAt = "2026-08-20T00:00:00Z",
    version = 1,
)

class VoiceServiceTest {

    private lateinit var server: MockWebServer
    private lateinit var api: SupabaseVoiceCaptureApi

    @Before
    fun setUp() {
        server = MockWebServer()
        server.start()
        api = SupabaseVoiceCaptureApi(
            config = PublicSupabaseConfig(server.url("/").toString(), "anon-key"),
            zoneIdProvider = { java.time.ZoneId.of("UTC") },
        )
    }

    @After
    fun tearDown() {
        server.shutdown()
    }

    private fun enqueueSuccess(body: String) {
        server.enqueue(
            MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(body)
        )
    }

    private fun createResponseBody(): String {
        // "tomorrow"/"today" grounded server-side relative to recording start.
        return """
        {
          "transcript": "Buy milk tomorrow and call doctor at 3pm",
          "mode": "create",
          "drafts": [
            {"title": "Buy milk", "description": null, "priority": 4,
             "plan_date": "2026-08-22", "plan_time": null, "plan_type": null},
            {"title": "Call doctor", "description": null, "priority": 2,
             "plan_date": "2026-08-21", "plan_time": "15:00:00", "plan_type": null}
          ],
          "edit": null
        }
        """.trimIndent()
    }

    private fun options(
        recordingStartTime: String = "2026-08-21T10:00:00Z",
        focusedTask: Task? = null,
        existingDrafts: List<DraftTask>? = null,
    ) = VoiceCaptureRequestOptions(
        audioWav = createWavHeader(32000),
        recordingStartTime = java.time.Instant.parse(recordingStartTime),
        timezone = "UTC",
        focusedTask = focusedTask,
        existingDrafts = existingDrafts,
        effectiveDefaultTimedPlanType = TimedPlanType.INSTANT,
    )

    // ---- request contract ----

    @Test
    fun `sends multipart with bearer auth, wav file, timezone and anchors`() = runTest {
        enqueueSuccess(createResponseBody())

        api.sendVoiceCapture(session(), options())

        val recorded = server.takeRequest(5, TimeUnit.SECONDS)!!
        assertEquals("Bearer test-access-token", recorded.getHeader("Authorization"))
        val body = recorded.body.readUtf8()
        assertTrue(body.contains("name=\"audio\""))
        assertTrue(body.contains("filename=\"audio.wav\""))
        assertTrue(body.contains("name=\"timezone\""))
        assertTrue(body.contains("UTC"))
        // Recording start time is sent verbatim as ISO timestamp.
        assertTrue(body.contains("2026-08-21T10:00:00Z"))
    }

    @Test
    fun `relative dates stay anchored to recording start even across midnight`() = runTest {
        // Recording began just before midnight UTC; processing happens after midnight.
        enqueueSuccess(createResponseBody())

        val result = api.sendVoiceCapture(
            session(),
            options(
                recordingStartTime = "2026-08-21T23:59:30Z",
                focusedTask = null,
            ),
        )

        val recorded = server.takeRequest(5, TimeUnit.SECONDS)!!
        val body = recorded.body.readUtf8()
        // The pre-midnight recording start reaches the boundary unchanged.
        assertTrue(body.contains("name=\"recording_start_time\""))
        assertTrue(body.contains("2026-08-21T23:59:30Z"))

        // Server-grounded dates land in drafts untouched by any local re-grounding:
        // "tomorrow" relative to the 23:59:30Z anchor is 2026-08-22.
        assertEquals(Plan.DateOnly(date = "2026-08-22"), result.drafts[0].plan)
    }

    @Test
    fun `includes focused_task context json when editing`() = runTest {
        enqueueSuccess(
            """
            {"transcript": "Move to tomorrow at 11am", "mode": "edit", "drafts": [],
             "edit": {"title": null, "description": null, "priority": null,
                      "plan_date": "2026-08-22", "plan_time": "11:00:00",
                      "plan_type": null, "clear_plan": null}}
            """.trimIndent()
        )

        val focused = task(plan = Plan.Floating(date = "2026-08-20", time = "10:00"))
        api.sendVoiceCapture(session(), options(focusedTask = focused))

        val recorded = server.takeRequest(5, TimeUnit.SECONDS)!!
        val body = recorded.body.readUtf8()
        assertTrue(body.contains("name=\"focused_task\""))
        assertTrue(body.contains("\"id\":\"${focused.id}\""))
        assertTrue(body.contains("\"title\":\"Existing Task\""))
        assertTrue(body.contains("\"priority\":3"))
        // Existing timed type travels to the boundary so the extractor sees it.
        assertTrue(body.contains("\"type\":\"floating\""))
    }

    @Test
    fun `includes existing_drafts summary when correcting`() = runTest {
        enqueueSuccess(createResponseBody())

        val drafts = listOf(
            DraftTask(
                id = "draft-1", title = "Buy milk", description = null, priority = 4,
                plan = Plan.DateOnly(date = "2026-08-22"),
            ),
            DraftTask(
                id = "draft-2", title = "Go to pool", description = null, priority = 4,
                plan = Plan.DateOnly(date = "2026-08-25"),
            ),
        )
        api.sendVoiceCapture(session(), options(existingDrafts = drafts))

        val recorded = server.takeRequest(5, TimeUnit.SECONDS)!!
        val body = recorded.body.readUtf8()
        assertTrue(body.contains("name=\"existing_drafts\""))
        assertTrue(body.contains("\"target_draft_index\":0"))
        assertTrue(body.contains("\"title\":\"Go to pool\""))
        assertTrue(body.contains("\"plan_date\":\"2026-08-25\""))
        // Untimed drafts omit plan_type entirely, like JSON.stringify dropping undefined.
        assertFalse(body.contains("\"plan_type\""))
    }

    // ---- successful response mapping ----

    @Test
    fun `maps successful multi-draft create response`() = runTest {
        enqueueSuccess(createResponseBody())

        val result = api.sendVoiceCapture(session(), options())

        assertEquals("Buy milk tomorrow and call doctor at 3pm", result.transcript)
        assertEquals(VoiceCaptureMode.CREATE, result.mode)
        assertEquals(2, result.drafts.size)
        assertEquals("Buy milk", result.drafts[0].title)
        assertEquals(Plan.DateOnly(date = "2026-08-22"), result.drafts[0].plan)
        assertEquals("Call doctor", result.drafts[1].title)
        assertEquals(2, result.drafts[1].priority)
        assertEquals(Plan.Instant(at = "2026-08-21T15:00:00Z"), result.drafts[1].plan)
        assertNull(result.editProposal)
    }

    @Test
    fun `edit mode preserves existing timed type unless speech changed it`() = runTest {
        enqueueSuccess(
            """
            {"transcript": "Move to tomorrow at 11am", "mode": "edit", "drafts": [],
             "edit": {"title": null, "description": null, "priority": null,
                      "plan_date": "2026-08-22", "plan_time": "11:00:00",
                      "plan_type": null, "clear_plan": null}}
            """.trimIndent()
        )

        val focused = task(plan = Plan.Floating(date = "2026-08-20", time = "10:00"))
        val result = api.sendVoiceCapture(session(), options(focusedTask = focused))

        assertEquals(VoiceCaptureMode.EDIT, result.mode)
        assertEquals(1, result.drafts.size)
        val editDraft = result.drafts[0]
        assertNotNull(result.editProposal)
        assertEquals(focused.id, editDraft.originalTaskId)
        assertEquals("Existing Task", editDraft.title) // title unchanged
        assertEquals("Original description", editDraft.description)
        assertEquals(3, editDraft.priority)
        // Preserved floating!
        assertEquals(Plan.Floating(date = "2026-08-22", time = "11:00"), editDraft.plan)
    }

    @Test
    fun `edit mode without edit payload mirrors the focused task`() = runTest {
        enqueueSuccess(
            """{"transcript": "Hmm", "mode": "edit", "drafts": [], "edit": null}"""
        )

        val focused = task()
        val result = api.sendVoiceCapture(session(), options(focusedTask = focused))

        val editDraft = result.drafts.single()
        assertEquals(focused.id, editDraft.originalTaskId)
        assertEquals(focused.title, editDraft.title)
        assertEquals(focused.description, editDraft.description)
        assertEquals(focused.priority, editDraft.priority)
        assertEquals(focused.plan, editDraft.plan)
        assertNull(editDraft.validationError)
    }

    @Test
    fun `correction merges updates into existing drafts by index and appends new ones`() = runTest {
        enqueueSuccess(
            """
            {"transcript": "No, buy milk today, not tomorrow", "mode": "create",
             "drafts": [
               {"target_draft_index": 0, "title": "Buy milk", "description": null,
                "priority": 4, "plan_date": "2026-08-21", "plan_time": null, "plan_type": null}
             ], "edit": null}
            """.trimIndent()
        )

        val existingDrafts = listOf(
            DraftTask(
                id = "draft-1", title = "Buy milk", description = null, priority = 4,
                plan = Plan.DateOnly(date = "2026-08-22"),
            ),
            DraftTask(
                id = "draft-2", title = "Go to pool", description = null, priority = 4,
                plan = Plan.DateOnly(date = "2026-08-25"),
            ),
        )
        val result = api.sendVoiceCapture(
            session(),
            options(existingDrafts = existingDrafts),
        )

        assertEquals(VoiceCaptureMode.CREATE, result.mode)
        assertEquals(2, result.drafts.size)
        assertEquals(Plan.DateOnly(date = "2026-08-21"), result.drafts[0].plan) // updated!
        assertEquals("Go to pool", result.drafts[1].title)
        assertEquals(Plan.DateOnly(date = "2026-08-25"), result.drafts[1].plan) // untouched!
    }

    @Test
    fun `clear_plan removes the plan from an edit draft`() = runTest {
        enqueueSuccess(
            """
            {"transcript": "Remove the date", "mode": "edit", "drafts": [],
             "edit": {"title": null, "description": null, "priority": null,
                      "plan_date": null, "plan_time": null, "plan_type": null,
                      "clear_plan": true}}
            """.trimIndent()
        )

        val result = api.sendVoiceCapture(session(), options(focusedTask = task()))

        assertNull(result.drafts.single().plan)
    }

    @Test
    fun `explicit instant type without clock time in edit flags validation error`() = runTest {
        enqueueSuccess(
            """
            {"transcript": "Make it instant tomorrow", "mode": "edit", "drafts": [],
             "edit": {"title": null, "description": null, "priority": null,
                      "plan_date": "2026-08-22", "plan_time": null,
                      "plan_type": "instant", "clear_plan": null}}
            """.trimIndent()
        )

        val result = api.sendVoiceCapture(session(), options(focusedTask = task()))

        val editDraft = result.drafts.single()
        assertEquals(
            "An explicit Instant or Floating plan requires a clock time.",
            editDraft.validationError,
        )
        assertEquals(Plan.DateOnly(date = "2026-08-22"), editDraft.plan)
    }

    // ---- error mapping ----

    @Test
    fun `maps 429 allowance errors with earliest retry metadata`() = runTest {
        server.enqueue(
            MockResponse().setResponseCode(429)
                .setHeader("Content-Type", "application/json")
                .setBody(
                    """
                    {"error": "Rate limit exceeded: maximum 3 requests per minute.",
                     "code": "rate_limit_minute",
                     "earliest_retry_at": "2026-08-21T10:01:00Z",
                     "retry_after_seconds": 45}
                    """.trimIndent()
                )
        )

        val error = runCatching { api.sendVoiceCapture(session(), options()) }
            .exceptionOrNull()

        assertTrue(error is VoiceError)
        error as VoiceError
        assertEquals(429, error.status)
        assertEquals("rate_limit_minute", error.code)
        assertEquals("2026-08-21T10:01:00Z", error.earliestRetryAt)
        assertEquals(45, error.retryAfterSeconds)
        assertFalse(error.isNetworkError)
    }

    @Test
    fun `maps 503 circuit breaker errors`() = runTest {
        server.enqueue(
            MockResponse().setResponseCode(503)
                .setHeader("Content-Type", "application/json")
                .setBody(
                    """
                    {"error": "Voice capture is temporarily unavailable. Please try again later.",
                     "code": "circuit_breaker_daily"}
                    """.trimIndent()
                )
        )

        val error = runCatching { api.sendVoiceCapture(session(), options()) }
            .exceptionOrNull() as VoiceError

        assertEquals(503, error.status)
        assertEquals("circuit_breaker_daily", error.code)
    }

    @Test
    fun `maps 502 provider errors`() = runTest {
        server.enqueue(
            MockResponse().setResponseCode(502)
                .setHeader("Content-Type", "application/json")
                .setBody("""{"error": "Voice processing failed. Please try again.", "code": "provider_error"}""")
        )

        val error = runCatching { api.sendVoiceCapture(session(), options()) }
            .exceptionOrNull() as VoiceError

        assertEquals(502, error.status)
        assertEquals("provider_error", error.code)
        assertEquals("Voice processing failed. Please try again.", error.message)
    }

    @Test
    fun `flags network failures so the recording is preserved`() = runTest {
        server.shutdown()
        val deadApi = SupabaseVoiceCaptureApi(
            config = PublicSupabaseConfig("http://127.0.0.1:1/", "anon-key"),
        )

        val error = runCatching {
            deadApi.sendVoiceCapture(session(), options())
        }.exceptionOrNull()

        assertTrue(error is VoiceError)
        error as VoiceError
        assertEquals(0, error.status)
        assertEquals("network_error", error.code)
        assertTrue(error.isNetworkError)
    }
}
